### PR TITLE
various code fixes

### PR DESCRIPTION
**Pull request by nschloe**
From [the coverity report](https://scan.coverity.com/projects/emulationstation?tab=overview).

### CODE REVIEW

File: es-app/src/VolumeControl.cpp [hunk 1]
```diff
@@ -32,7 +32,15 @@ VolumeControl::VolumeControl()
...
+VolumeControl::VolumeControl(const VolumeControl & right):
+  originalVolume(0), internalVolume(0)
+#if defined (__APPLE__)
+    #error TODO: Not implemented for MacOS yet!!!
+#elif defined(__linux__)
+    , mixerIndex(0), mixerHandle(nullptr), mixerElem(nullptr), mixerSelemId(nullptr)
+#elif defined(WIN32) || defined(_WIN32)
+	, mixerHandle(nullptr), endpointVolume(nullptr)
+#endif
 {
```
Comment (joolswills) [inline es-app/src/VolumeControl.cpp]: wrong indentation - es uses tabs

**Comment by joolswills** [comment]
i've not had time to check the report yet - any comments on this @jrassa / @pjft ?

File: es-core/src/animations/LambdaAnimation.h [hunk 2]
```diff
@@ -8,6 +8,8 @@ class LambdaAnimation : public Animation
 public:
 	LambdaAnimation(const std::function<void(float t)>& func, int duration) : mFunction(func), mDuration(duration) {}
 
+  virtual ~LambdaAnimation() = default;
+
```
Comment (joolswills) [inline es-core/src/animations/LambdaAnimation.h]: wrong indentation.

File: es-app/src/VolumeControl.cpp (hunk 1 above)
Comment (joolswills) [inline es-app/src/VolumeControl.cpp]: (The file has some other areas with bad indentation - feel free to correct the whitespace in a separate commit or I will do it after the PR).

**Comment by pjft** [comment]
I was not aware of this report, but have requested access to it and will take a look.

I have no significant concerns with these changes in principle.

**Comment by zigurana** [comment]
@pjft, same here. Is this report not based on the Aloshi-branch? If it does not take into account our latest work, I think this will be of limited use for us?

**Comment by nschloe** [comment]
The report is based on the `master` branch.

File: es-core/src/Window.cpp [hunk 3]
```diff
@@ -335,8 +335,8 @@ void Window::setHelpPrompts(const std::vector<HelpPrompt>& prompts, const HelpSt
 				// yes, it has!
 
 				// can we combine? (dpad only)
-				if((it->first == "up/down" && addPrompts.at(mappedTo->second).first == "left/right") ||
-					(it->first == "left/right" && addPrompts.at(mappedTo->second).first == "up/down"))
+				if((strcmp(it->first, "up/down") == 0 && addPrompts.at(mappedTo->second).first == "left/right") ||
```
Comment (zigurana) [inline es-core/src/Window.cpp]: What is the use of this change? If I understood correctly, `string == "Foo"` is exactly the same thing as `string.compare("Foo") == 0`. Also, strcmp is mostly used for C-style strings, right?

Comment (nschloe) [inline es-core/src/Window.cpp]: If I'm not mistaken, all quantities are `(const) char *` here, in which case `it->first == "up/down"` doesn't do what you'd expect: It really compares the addresses of the `char *`. `strcmp` actually performs string comparison.

Comment (pjft) [inline es-core/src/Window.cpp]: I was looking into this exact same set of changes.

It seems the HelpPrompts are a pair of standard C strings (const char*).

Read: https://www.daniweb.com/programming/software-development/threads/78711/why-strcmp-when-can-use-operator

As such, I did not object to the changes here. I know C is a lot trickier in this kind of things, and I am far from up to date on that.

**Comment by zigurana** [comment]
@nschloe : yes, the masterbranch of https://github.com/Aloshi/EmulationStation, not of https://github.com/RetroPie/EmulationStation, right?

**Comment by nschloe** [comment]
@zigurana Nope, the retropie version. I saw just now that the meta data was still wrong (fixed now).

File: es-core/src/Window.cpp (hunk 3 above)
Comment (zigurana) [inline es-core/src/Window.cpp]: @nschloe , @pjft : thanks for the explanation, but that would indicate that this code would never have worked properly, or not?

**Comment by zigurana** [comment]
@nschloe , thanks for adding me to the project, looks interesting!

File: es-core/src/Window.cpp (hunk 3 above)
Comment (pjft) [inline es-core/src/Window.cpp]: I was hoping @nschloe would answer first :D I've learned time and time again not to question C, as I am never better off after it. Just the past hour I've been trying to understand why a simple division assignment to an unsigned char wasn't working as intended, in EmulationStation... 

So yes, in principle your assessment would be correct. I wonder if the compiler does some sort of smart optimization or conversion to make it work - or, if as you say, it actually never really worked properly. 

Then again, maybe I am happier not knowing the answer :)

Comment (nschloe) [inline es-core/src/Window.cpp]: It never worked. MWE:
```
#include <iostream>
#include <string.h>

int main() {
  char c[] = "abc";
  if (c == "abc") {
    std::cout << "okay 1" << std::endl;
  }
  if (strcmp(c, "abc") == 0) {
    std::cout << "okay 2" << std::endl;
  }
}
```
prints
```
okay 2
```

**Comment by nschloe** [comment]
So, can this be merged?

File: es-core/src/Window.cpp (hunk 3 above)
Comment (pjft) [inline es-core/src/Window.cpp]: Question: shouldn't the addPrompts comparison also be with strcmp then in these two lines?

addPrompts.at(mappedTo->second).first == "left/right")

File: es-core/src/components/ComponentList.cpp [hunk 4]
```diff
@@ -319,7 +319,7 @@ std::vector<HelpPrompt> ComponentList::getHelpPrompts()
 		bool addMovePrompt = true;
 		for(auto it = prompts.begin(); it != prompts.end(); it++)
 		{
-			if(it->first == "up/down" || it->first == "up/down/left/right")
+			if(strcmp(it->first, "up/down") ==0 || strcmp(it->first, "up/down/left/right") == 0)
```
Comment (pjft) [inline es-core/src/components/ComponentList.cpp]: Apologies for the nitpick: missing space between == and 0, if you have the chance. :/

**Comment by pjft** [comment]
Added a question and a comment, but otherwise - assuming it compiles - I have no objections.

Thanks.

**Comment by nschloe** [comment]
@pjft Done.

**Comment by pjft** [comment]
Thanks. 

Just compiling and giving it a whirl. Will report back.

**Comment by joolswills** [comment]
Please can you squash the additional changes. Thanks.

**Comment by pjft** [comment]
Tested. Compiles. No functional changes that I am aware of though - which is good.

On the other hand, I do want to look into the string comparison piece and understand whether the compiler couldn't be doing some optimization on their end, as some code that shouldn't have been working seems to have been working. I know - I find it weird as well.

Anyway, thanks for the changes and bringing these forward!

**Comment by joolswills** [comment]
Thanks.

**Comment by joolswills** [comment]
Would you mind squashing the additional fixes so it's nicely grouped etc?

File: es-core/src/guis/GuiInputConfig.cpp [hunk 5]
```diff
@@ -396,7 +396,7 @@ bool GuiInputConfig::assign(Input input, int inputId)
 
 	// if this input is mapped to something other than "nothing" or the current row, error
 	// (if it's the same as what it was before, allow it)
-	if(mTargetConfig->getMappedTo(input).size() > 0 && !mTargetConfig->isMappedTo(inputName[inputId], input) && inputName[inputId] != "HotKeyEnable")
+	if(mTargetConfig->getMappedTo(input).size() > 0 && !mTargetConfig->isMappedTo(inputName[inputId], input) && strcmp(inputName[inputId], "HotKeyEnable") != 0)
```
Comment (joolswills) [inline es-core/src/guis/GuiInputConfig.cpp]: Based on the comments the old code with the hotkey compare shouldn't work? Except it does? Interested to know why. Compiler being clever?

**Comment by nschloe** [comment]
Squashing as in squashing commits? I guess this is best done in GitHub's interface by whoever merges it.

**Comment by hex007** [comment]
@nschloe You should be able to do it and force push to have a single commit. Considering you have 6 commits:

```
git reset HEAD~6
git commit
```

**Comment by nschloe** [comment]
@hex007 Done.

IMHO, squashing should really be the responsibility of upstream, particularly since [it's so easy to do in GitHub](https://github.com/blog/2141-squash-your-commits).

File: es-core/src/Window.cpp [hunk 6]
```diff
@@ -335,8 +335,8 @@ void Window::setHelpPrompts(const std::vector<HelpPrompt>& prompts, const HelpSt
 				// yes, it has!
 
 				// can we combine? (dpad only)
-				if((it->first == "up/down" && addPrompts.at(mappedTo->second).first == "left/right") ||
-					(it->first == "left/right" && addPrompts.at(mappedTo->second).first == "up/down"))
+				if((strcmp(it->first, "up/down") == 0 && strcmp(addPrompts.at(mappedTo->second).first, "left/right")) ||
```
Comment (hex007) [inline es-core/src/Window.cpp]: Is strcmp the way to go? it->first is probably string and not char*

Comment (nschloe) [inline es-core/src/Window.cpp]: Coverity suggested otherwise.

**Comment by pjft** [comment]
Correct. It is unfortunately char* .

**Comment by joolswills** [comment]
Sometimes commits are best squashed into logical groups. But anyway, I prefer it with a merge commit rather than a single squashed commit via github interface.

**Comment by joolswills** [comment]
@hex007 resetting is not a good way of squashing. You lose the commit messages. It's done via git rebase -i HEAD~6 for the last 6 commits etc.

**Pull request merged**